### DISCUSSION TITLE
image processor for file compression

### DISCUSSION THREAD

**Discussion by CuddleBear92** [Difegue/LANraragi#332] [Ideas]
The addition of an image processor to compress filesizes would really make sense. 
To be able to for example convert any image format into .webp at a set quality % would make sense.
Allowing users to say set the quality to 90% or even 80% unless you have a neat way to crush the sizes without loss. 

Personally have not seen much to losses at 80% in ComicRack while crushing the files down to under half the size, more with black and white. Something similar in LANraragi makes sense i think as people have larger and larger collections. 

Webp shouldnt cause issues in this as browsers support it across the board. 
But having the choice of multiple filetypes makes sense as i guess people might have usecases to convert png files to say other files types aswell.

webp, heif and AVIF is prob the ways to go in the future for smaller sizes.
Having some sort of metadata to not process the same files again would make sense, esp when quality % is used.

**Comment by Difegue** [comment]
LRR currently bundles imagemagick with webp support -- I think HEIF support is in too but don't quote me on that, I don't know if I added libheif to the mix.  

Viewer support is up to the browsers as you said, so this feature is entirely within the realm of possibility.  

I'd write it as a script plugin that'd do the following:

1. Extract specified archive (by id) to a temp folder w. libarchive (internal API already exists in LRR for that)
2. Use magick to convert all files to the specified format and quality%
3. Repack archive w. libarchive (need to write some code here as well, there's no internal API for repacking in LRR)
4. Compute ID of the repacked archive, update Redis records to replace the old id with the new
5. Return new id as script result to inform the user and probably wipe the search cache while you're at it
